### ISSUE TITLE
MNIST training example on GoTorch and LibTorch can not convergence to the same accuracy

### ISSUE THREAD

**Issue by Yancey1989** [wangkuiyi/gotorch#113]
LibTorch [MNIST example](https://github.com/wangkuiyi/gotorch/blob/develop/benchmark/main.cpp) got the loss **0.0269** after 5 epochs:

``` text
Epoch 0, Loss: 0.1280
Epoch 1, Loss: 0.0659
Epoch 2, Loss: 0.0396
Epoch 3, Loss: 0.0304
Epoch 4, Loss: 0.0269
```

GoTorch [MNIST example](https://github.com/wangkuiyi/gotorch/blob/develop/mnist_test.go) got the loss **1.4148** after 5 epochs:

``` text
2020/08/12 22:37:41 Epoch: 0, Loss: 4.8264
2020/08/12 22:37:46 Epoch: 1, Loss: 5.9624
2020/08/12 22:37:52 Epoch: 2, Loss: 2.4493
2020/08/12 22:37:58 Epoch: 3, Loss: 0.9619
2020/08/12 22:38:04 Epoch: 4, Loss: 1.4148
```

**Comment by Yancey1989** [comment]
After merged #115 and #116 

Go version got a similar  loss, but  there is also an overlap  between C++ and Go version

``` text
2020/08/13 00:40:43 Epoch: 0, Loss: 0.2228
2020/08/13 00:40:48 Epoch: 1, Loss: 0.1510
2020/08/13 00:40:53 Epoch: 2, Loss: 0.0509
2020/08/13 00:40:58 Epoch: 3, Loss: 0.0622
2020/08/13 00:41:04 Epoch: 4, Loss: 0.0786
```

**Comment by Yancey1989** [comment]
I tried to compare the weight of the FC layer between  C++ and Go version,  the weight of `fc1` is the same, but different on `fc2`.

It seems `torch::manual_seed`  can not fix the random seed,  I did two experiments just like the following:

- Case1: set the random seed on top of the program

``` c++
torch::manual_seed(1);
auto t1 = torch::empty({2, 3});
auto t2 = torch::empty({2, 3});
torch::nn::init::kaiming_uniform_(t1, std::sqrt(5));
torch::nn::init::kaiming_uniform_(t2, std::sqrt(5));
```

```  bash
➜  gotorch git:(develop) ✗ ./benchmark/mnist
 0.2975 -0.2548 -0.1119
 0.2710 -0.5435  0.3462
[ CPUFloatType{2,3} ]
-0.1188  0.2937  0.0803
-0.0707  0.1601  0.0285
[ CPUFloatType{2,3} ]
➜  gotorch git:(develop) ✗ ./benchmark/mnist
 0.2975 -0.2548 -0.1119
 0.2710 -0.5435  0.3462
[ CPUFloatType{2,3} ]
-0.1188  0.2937  0.0803
-0.0707  0.1601  0.0285
[ CPUFloatType{2,3} ]
```

conclusion: `t2` got a different value, and it's stable on two executions.

- Case2:  Set random seed before each kaiming initialization


``` c++
torch::manual_seed(1);
auto t1 = torch::empty({2, 3});
auto t2 = torch::empty({2, 3});
torch::nn::init::kaiming_uniform_(t1, std::sqrt(5));
torch::manual_seed(1);
torch::nn::init::kaiming_uniform_(t2, std::sqrt(5));
```

```  bash
➜  gotorch git:(develop) ✗ ./benchmark/mnist
 0.2975 -0.2548 -0.1119
 0.2710 -0.5435  0.3462
[ CPUFloatType{2,3} ]
-0.1188  0.2937  0.0803
-0.0707  0.1601  0.0285
[ CPUFloatType{2,3} ]
➜  gotorch git:(develop) ✗ ./benchmark/mnist
 0.2975 -0.2548 -0.1119
 0.2710 -0.5435  0.3462
[ CPUFloatType{2,3} ]
-0.1188  0.2937  0.0803
-0.0707  0.1601  0.0285
[ CPUFloatType{2,3} ]
```

Con:  `t1` and `t2` got the same value, and it's stable.
